### DISCUSSION TITLE
Backup Mikrotik

### DISCUSSION THREAD

**Discussion by SerhiiZahuba** [DanSheps/netbox-config-backup#40] [Q&A]
I have a question, maybe you can give me some advice. For device backups, I use https://cbackup.readthedocs.io/ the project has not been developed for a long time, but it suits me almost completely. It uses templates to connect via ssh\telnet to the device and makes a backup. Mikrotik recently updated the ssh version in its firmware and this tool stopped working - it connects to the router but sees a blank page. Napalm-ros can't show device config. Is it possible to adapt your tool to Mikrotik backups?

**Comment by DanSheps** [comment]
This relies on Napalm under the hood, so likely something broke the napalm portion of this.

I don't have a Microtik device, so if you could do the following from the netbox shell:

```python
>>> from netbox_config_backup.tasks import napalm_init
>>> device = Device.objects.get(name="<device>")
>>> d = napalm_init(device, device.primary_ip)
>>> configs = d.get_config()
>>> configs.keys()
>>> len(configs.get('startup'))
>>> len(configs.get('running'))
>>> len(configs.get('candidate'))
```
Paste back any of the output if you could (you can sanitize your device name, nothing else should leak sensistive information)

**Comment by SerhiiZahuba** [comment]
>>> d = napalm_init(device, device.primary_ip)
Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.9/site-packages/netbox_config_backup/tasks.py", line 58, in napalm_init
    driver = napalm.get_network_driver(device.platform.napalm_driver)
  File "/opt/netbox/venv/lib/python3.9/site-packages/napalm/base/__init__.py", line 97, in get_network_driver
    raise ModuleImportError(
napalm.base.exceptions.ModuleImportError: Cannot import "napalm-ros". Is the library installed?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/opt/netbox/venv/lib/python3.9/site-packages/netbox_config_backup/tasks.py", line 60, in napalm_init
    raise ServiceUnavailable("NAPALM driver for platform {} not found: {}.".format(
netbox.api.exceptions.ServiceUnavailable: NAPALM driver for platform ROS (mikrotik) not found: napalm-ros.



now exiting InteractiveConsole...
(venv) [root@Inventory netbox]# pip list module | grep ros
librouteros                   3.2.0
napalm-ros                    1.0.1
(venv) [root@Inventory netbox]#

**Reply by DanSheps** [reply]
What if you just run the python console and try to initiate napalm with the ros driver from there?

**Comment by SerhiiZahuba** [comment]
Sorry, can you describe in detail what needs to be done?

**Reply by DanSheps** [reply]
You need to test napalm outside of NetBox but in your venv

**Reply by SerhiiZahuba** [reply]
![image](https://user-images.githubusercontent.com/36307887/166105178-f10294e9-4c00-4420-9739-1be2b7b22523.png)

I tried to connect via python, did everything with https://github.com/joncutrer/python-tutorials/blob/master/net/automation/RouterOS-Automation-with-NAPALM.ipynb

The developers of this module wrote that they will add the output of information about the config in the next release

https://github.com/napalm-automation-community/napalm-ros/issues/92#issuecomment-1109078540

**Comment by SerhiiZahuba** [comment]
Hello. If you still support this project. Napalm started to support config output for Mikrotik. From my env I can get the config for the device

print(device.get_config())

If you have the opportunity, please add support for Mikrotik.


![image](https://user-images.githubusercontent.com/36307887/236817873-407c5ad8-0a8d-4b0b-806f-6f10b39e347a.png)
![image](https://user-images.githubusercontent.com/36307887/236817974-c855f226-e6b0-4b1c-8dcc-4d693eee91c3.png)

**Reply by DanSheps** [reply]
Please open an issue and I will see what I can do.
